implement M3, M4 & M5 spindle control commands

Bob-the-Kuhn opened the pull request:
all code changes are in Marlin_main.cpp

none of the existing code was changed

section added to config_adv to enable/disable the new commands & to explain the hardware changes needed to properly connect to the spindle controller

There's lots of competition for the PWM pins so added a section in sanity.h to check that the selected pin doesn't conflict with other selections.

Bob-the-Kuhn committed:
implement M3, M4 & M5 spindle control commands

This is a combination of 5 commits.

The major changes since the first commit are:
 *  synchronized with the G0, G1, G2 & G3 movement commands
 *  added section to config_adv enable/disable the commands
 *  added section to sanitycheck.h to check for resource conflicts
 *  gave the user more control over the parameters used by the commands
 *  added power up/down delays to allow spindle speed to stabilize before next movement command
 *  lots of changes to make it follow Marlin conventions and good coding practices

The following are the original commit messages.

================================================================

The first commit's message is:

implement M3, M4 & M5 spindle control commands

all code changes are in Marlin_main.cpp

none of the existing code was changed

section added to config_adv to enable/disable the new commands & to
explain the hardware changes needed to properly connect to the spindle
controller

There's lots of competition for the PWM pins so added a section in
sanity.h to check that the selected pin doesn't conflict with other
selections.

================================================================

This is the commit message #2:

added M3, M4 & M5 to the list of implemented commands

================================================================

This is the commit message #3:

bug fix, more generic, style/convention updates

1) fixed a bug when setting the maximum RPM

2) made it more generic - allowed the user to set the RPM to PWM
conversion constants in config_adv

3) lots of changes that Scott suggested

Changed “SPINDLE_ON_OFF” to “SPINDLE_ENABLE”

Changed “digitalWrite(” to “WRITE(“

Went to “#if ENABLED(SPINDLE_SPEED)” and ““#if ENABLED(SPINDLE_ENABLE)”

Removed the debug prints

Indented all the comments by 2 spaces

Got rid of the extra “pinMode” commands

Went to “gcode_M3_M4”

Went to “elseif” in sanitycheck.h

================================================================

This is the commit message #4:

add power up/down delays,  bug fix & add error msg

1) added user defined power up and power down delays to allow the
spindle to stabilize

2) bug fix - synchronized M3, M4 & M5 to prevent out of sequence command
execution

3) sanitycheck mod to add "SPINDLE_ENABLE requires SPINDLE_ENABLE_PIN."
message

================================================================

This is the commit message #5:

cleanup

cleanup indentation & formatting

removed "MOTOR_CURRENT_PWM" from sanity checks

converted delay macros to functions

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4279,6 +4282,144 @@ inline void gcode_G92() {
 #endif // ULTIPANEL || EMERGENCY_PARSER
 
+
+#if defined(SPINDLE_ON_OFF) && SPINDLE_ON_OFF_PIN != -1
```
This should be…

``` cpp
#if ENABLED(SPINDLE_ON_OFF) && PIN_EXISTS(SPINDLE_ON_OFF)
```

…however if the sanity check already checks for the spindle pin, then it only needs…

``` cpp
#if ENABLED(SPINDLE_ON_OFF)
```

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4279,6 +4282,144 @@ inline void gcode_G92() {
+
+#if defined(SPINDLE_ON_OFF) && SPINDLE_ON_OFF_PIN != -1
+ #define SPINDLE_ENABLED
+ #if defined (SPINDLE_SPEED) && SPINDLE_SPEED_PIN != -1
```
And again…

``` cpp
#if ENABLED(SPINDLE_SPEED) && PIN_EXISTS(SPINDLE_SPEED)
```

…but if the sanity check already looks for the pin…

``` cpp
#if ENABLED(SPINDLE_SPEED)
```

…is all you need.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4279,6 +4282,144 @@ inline void gcode_G92() {
+  #endif
+#endif
+
+ #ifdef SPINDLE_ENABLED
```
``` cpp
#if ENABLED(SPINDLE_ENABLED)
```

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4279,6 +4282,144 @@ inline void gcode_G92() {
+ 
+// PWM duty cycle goes from 0 (off) to 255 (always on) 
+ 
+ inline void gcode_M3() {
```
The condition (`#if ENABLED(SPINDLE_SPEED_ENABLED)`) should go around the function definitions, so the functions won't even be defined unless the option is set.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4279,6 +4282,144 @@ inline void gcode_G92() {
+#if defined(SPINDLE_ON_OFF) && SPINDLE_ON_OFF_PIN != -1
+ #define SPINDLE_ENABLED
+ #if defined (SPINDLE_SPEED) && SPINDLE_SPEED_PIN != -1
+    #define SPINDLE_SPEED_ENABLED
```
Conditional defines like these are better placed in `Conditionals.h`.

Although, actually you don't need to double up on the options, adding `SPINDLE_ENABLED` and `SPINDLE_SPEED_ENABLED`. Instead, just confirm in `SanityCheck.h` that the pins the features need are defined when enabled and throw an error if they are not. Then the code only needs to look for the principal options.

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,68 @@
+ #if !defined(SPINDLE_ON_OFF_INVERT) || !(SPINDLE_ON_OFF_INVERT == true || SPINDLE_ON_OFF_INVERT == false)
+   #error "SPINDLE_ON_OFF_INVERT missing or incorrectly defined"
+ #endif 
+ #if defined (SPINDLE_SPEED) && SPINDLE_SPEED_PIN != -1
```
The `#endif` / `#if` pairs can be condensed to `#elif`.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4279,6 +4282,144 @@ inline void gcode_G92() {
+  #ifdef SPINDLE_SPEED_ENABLED
+     uint8_t spindle_speed = code_seen('S') ? code_value_byte() : 0; 
+     if (spindle_speed == 0) { 
+         digitalWrite(SPINDLE_ON_OFF_PIN, !SPINDLE_ON_OFF_INVERT);  //turn spindle off (active low);
```
Indentation is 2 spaces.

Where possible (here for example), replace the `digitalWrite` function with the fastio equivalent `WRITE` macro.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4279,6 +4282,144 @@ inline void gcode_G92() {
+      else {
+         digitalWrite(SPINDLE_ON_OFF_PIN, SPINDLE_ON_OFF_INVERT); //turn spindle on (active low) if spindle speed option not enabled
+         pinMode(SPINDLE_SPEED_PIN, OUTPUT);
+         #if SPINDLE_SPEED_INVERT==true
```
For boolean settings option, no need to use `true` and `false`. Just `#define` to set true and leave undefined for false, then use `ENABLED(SPINDLE_SPEED_INVERT)` macro. This method still allows a Makefile to force an off setting by using values `false` or `0`.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -9421,6 +9576,19 @@ void setup() {
   servo_init();
 
+  #ifdef SPINDLE_ENABLED
+     pinMode(SPINDLE_ON_OFF_PIN, OUTPUT);
```
No need to do `pinMode(pin, OUTPUT)` here because `OUT_WRITE` will do that below.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4279,6 +4282,144 @@ inline void gcode_G92() {
+ * M5 turn off spindle
+ */
+inline void gcode_M5() { 
+  digitalWrite(SPINDLE_ON_OFF_PIN, !SPINDLE_ON_OFF_INVERT);  // turn spindle off
```
Better:

``` cpp
WRITE(SPINDLE_ON_OFF_PIN, !SPINDLE_ON_OFF_INVERT);
```

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,68 @@
+      #error "SPINDLE_SPEED PWM pin is used by MOTOR_CURRENT_PWM_E"
+    #endif 
+  #endif    
+  #if   defined(EXTRUDER_0_AUTO_FAN_PIN) && SPINDLE_SPEED_PIN == EXTRUDER_0_AUTO_FAN_PIN
```
``` cpp
#elif HAS_AUTO_FAN_0 && SPINDLE_SPEED_PIN == EXTRUDER_0_AUTO_FAN_PIN
```

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,68 @@
+  #if   defined(EXTRUDER_3_AUTO_FAN_PIN) && SPINDLE_SPEED_PIN == EXTRUDER_3_AUTO_FAN_PIN  
+     #error "SPINDLE_SPEED PWM pin is used by EXTRUDER_3_AUTO_FAN_PIN"
+  #endif  
+  #if   defined(FAN_PIN) && SPINDLE_SPEED_PIN == FAN_PIN 
```
``` cpp
#elif HAS_FAN0 && SPINDLE_SPEED_PIN == FAN_PIN
```

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,68 @@
+  #if   defined(FAN_PIN) && SPINDLE_SPEED_PIN == FAN_PIN 
+       #error "SPINDLE_SPEED PWM pin is used FAN_PIN "
+  #endif
+  #if    defined(FAN0_PIN) && SPINDLE_SPEED_PIN == FAN0_PIN 
```
There is no `FAN0_PIN`. `FAN_PIN` takes that slot.

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,68 @@
+  #if    defined(FAN0_PIN) && SPINDLE_SPEED_PIN == FAN0_PIN 
+       #error "SPINDLE_SPEED PWM pin is used FAN0_PIN "
+  #endif
+  #if   defined(FAN1_PIN) && SPINDLE_SPEED_PIN == FAN1_PIN
```
``` cpp
#elif HAS_FAN1 && SPINDLE_SPEED_PIN == FAN1_PIN
```

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,68 @@
+  #if   defined(FAN1_PIN) && SPINDLE_SPEED_PIN == FAN1_PIN
+       #error "SPINDLE_SPEED PWM pin is used FAN1_PIN "
+  #endif
+  #if   defined(FAN2_PIN) && SPINDLE_SPEED_PIN == FAN2_PIN
```
``` cpp
#elif HAS_FAN2 && SPINDLE_SPEED_PIN == FAN2_PIN
```

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,68 @@
+  #if   defined(FAN2_PIN) && SPINDLE_SPEED_PIN == FAN2_PIN
+     #error "SPINDLE_SPEED PWM pin is used FAN2_PIN "
+  #endif
+  #if defined(CONTROLLERFAN_PIN) && SPINDLE_SPEED_PIN == CONTROLLERFAN_PIN
```
``` cpp
#elif HAS_CONTROLLERFAN && SPINDLE_SPEED_PIN == CONTROLLERFAN_PIN
```

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,68 @@
+  #if   defined(EXTRUDER_0_AUTO_FAN_PIN) && SPINDLE_SPEED_PIN == EXTRUDER_0_AUTO_FAN_PIN
+     #error "SPINDLE_SPEED PWM pin is used by EXTRUDER_0_AUTO_FAN_PIN"
+  #endif  
+  #if   defined(EXTRUDER_1_AUTO_FAN_PIN) && SPINDLE_SPEED_PIN == EXTRUDER_1_AUTO_FAN_PIN 
```
``` cpp
#elif HAS_AUTO_FAN_1 && SPINDLE_SPEED_PIN == EXTRUDER_1_AUTO_FAN_PIN
```

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,68 @@
+  #if   defined(EXTRUDER_1_AUTO_FAN_PIN) && SPINDLE_SPEED_PIN == EXTRUDER_1_AUTO_FAN_PIN 
+	 #error "SPINDLE_SPEED PWM pin is used by EXTRUDER_1_AUTO_FAN_PIN"
+  #endif  
+  #if   defined(EXTRUDER_2_AUTO_FAN_PIN) && SPINDLE_SPEED_PIN == EXTRUDER_2_AUTO_FAN_PIN 
```
``` cpp
#elif HAS_AUTO_FAN_2 && SPINDLE_SPEED_PIN == EXTRUDER_2_AUTO_FAN_PIN
```

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,68 @@
+  #if   defined(EXTRUDER_2_AUTO_FAN_PIN) && SPINDLE_SPEED_PIN == EXTRUDER_2_AUTO_FAN_PIN 
+	 #error "SPINDLE_SPEED PWM pin is used by EXTRUDER_2_AUTO_FAN_PIN"
+  #endif  
+  #if   defined(EXTRUDER_3_AUTO_FAN_PIN) && SPINDLE_SPEED_PIN == EXTRUDER_3_AUTO_FAN_PIN  
```
``` cpp
#elif HAS_AUTO_FAN_3 && SPINDLE_SPEED_PIN == EXTRUDER_3_AUTO_FAN_PIN
```

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,68 @@
+     #error "Counter/Timer for SPINDLE_SPEED PWM pin is used by the servo system "
+  #endif
+  #if defined(HAS_MOTOR_CURRENT_PWM)
+    #if PIN_EXISTS(MOTOR_CURRENT_PWM_XY) &&  SPINDLE_SPEED_PIN == MOTOR_CURRENT_PWM_XY
```
``` cpp
#elif PIN_EXISTS(MOTOR_CURRENT_PWM_XY) && SPINDLE_SPEED_PIN == MOTOR_CURRENT_PWM_XY_PIN
```

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,68 @@
+    #if PIN_EXISTS(MOTOR_CURRENT_PWM_XY) &&  SPINDLE_SPEED_PIN == MOTOR_CURRENT_PWM_XY
+      #error "SPINDLE_SPEED PWM pin is used by MOTOR_CURRENT_PWM_XY"
+    #endif 
+    #if PIN_EXISTS(MOTOR_CURRENT_PWM_Z) &&  SPINDLE_SPEED_PIN == MOTOR_CURRENT_PWM_Z
```
``` cpp
#elif PIN_EXISTS(MOTOR_CURRENT_PWM_Z) && SPINDLE_SPEED_PIN == MOTOR_CURRENT_PWM_Z_PIN
```

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,68 @@
+    #if PIN_EXISTS(MOTOR_CURRENT_PWM_Z) &&  SPINDLE_SPEED_PIN == MOTOR_CURRENT_PWM_Z
+      #error "SPINDLE_SPEED PWM pin is used by MOTOR_CURRENT_PWM_Z"
+    #endif 
+    #if PIN_EXISTS(MOTOR_CURRENT_PWM_E) &&  SPINDLE_SPEED_PIN == MOTOR_CURRENT_PWM_E
```
``` cpp
#elif PIN_EXISTS(MOTOR_CURRENT_PWM_E) && SPINDLE_SPEED_PIN == MOTOR_CURRENT_PWM_E_PIN
```

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,68 @@
+  #if defined(X_MAX_PIN) &&  X_MAX_PIN == SPINDLE_SPEED_PIN && defined(USE_XMAX_PLUG)
+     #error "SPINDLE_SPEED pin is in use by X_MAX endstop"
+  #endif
+  #if defined(X_MIN) &&  X_MIN == SPINDLE_SPEED_PIN && defined(USE_XMIN_PLUG)
```
``` cpp
#elif HAS_X_MIN && SPINDLE_SPEED_PIN == X_MIN_PIN
```

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,68 @@
+  #if SPINDLE_SPEED_PIN == (4 || 11 || 12 || 13)
+     #error "Counter/Timer for SPINDLE_SPEED PWM pin is used by a system interrupt "
+  #endif
+  #if defined(X_MAX_PIN) &&  X_MAX_PIN == SPINDLE_SPEED_PIN && defined(USE_XMAX_PLUG)
```
``` cpp
#elif HAS_X_MAX && SPINDLE_SPEED_PIN == X_MAX_PIN
```

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,68 @@
+  #if defined(X_MIN) &&  X_MIN == SPINDLE_SPEED_PIN && defined(USE_XMIN_PLUG)
+     #error "SPINDLE_SPEED pin is in use by X_MIN endstop "
+  #endif
+  #if defined(Z_STEP_PIN) &&  Z_STEP_PIN == SPINDLE_SPEED_PIN
```
``` cpp
#elif HAS_Z_STEP && SPINDLE_SPEED_PIN == Z_STEP_PIN
```

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,68 @@
+  #if defined(Z_STEP_PIN) &&  Z_STEP_PIN == SPINDLE_SPEED_PIN
+     #error "SPINDLE_SPEED pin in use by Z_STEP "
+  #endif
+  #if defined(NUM_SERVOS) && (NUM_SERVOS > 0) &&  SPINDLE_SPEED_PIN == (2 || 3 || 5)
```
``` cpp
#elif HAS_SERVOS && (SPINDLE_SPEED_PIN == 2 || SPINDLE_SPEED_PIN == 3 || SPINDLE_SPEED_PIN == 5)
```

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,68 @@
+  #if !defined(SPINDLE_SPEED_INVERT) || !(SPINDLE_SPEED_INVERT == true || SPINDLE_SPEED_INVERT == false)
+    #error "SPINDLE_SPEED_INVERT missing or incorrectly defined"
+  #endif  
+  #if SPINDLE_SPEED_PIN == (4 || 11 || 12 || 13)
```
``` cpp
#elif SPINDLE_SPEED_PIN == 4 || SPINDLE_SPEED_PIN == 11 || SPINDLE_SPEED_PIN == 12 || SPINDLE_SPEED_PIN == 13
```

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,68 @@
+   #error "SPINDLE_ON_OFF_INVERT missing or incorrectly defined"
+ #endif 
+ #if defined (SPINDLE_SPEED) && SPINDLE_SPEED_PIN != -1
+  #if !defined(SPINDLE_SPEED_INVERT) || !(SPINDLE_SPEED_INVERT == true || SPINDLE_SPEED_INVERT == false)
```
``` cpp
#elif !defined(SPINDLE_SPEED_INVERT)
```

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,68 @@
 #endif
+
+
+#if defined( SPINDLE_ON_OFF) && SPINDLE_ON_OFF_PIN != -1 
```
``` cpp
#if ENABLED(SPINDLE_ON_OFF) && PIN_EXISTS(SPINDLE_ON_OFF)
```

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -7413,6 +7554,20 @@ void process_next_command() {
+     
+      #ifdef SPINDLE_ENABLED
+        case 3:
+          gcode_M3();
```
Following current conventions, this function should be called `gcode_M3_M4` instead.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -7413,6 +7554,20 @@ void process_next_command() {
           break;
       #endif // ULTIPANEL
+     
+      #ifdef SPINDLE_ENABLED
```
``` cpp
#if ENABLED(SPINDLE_ENABLED)
```

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4279,6 +4282,144 @@ inline void gcode_G92() {
+         } 
+     
+   #else
+    digitalWrite(SPINDLE_ON_OFF_PIN, SPINDLE_ON_OFF_INVERT);  //turn spindle on (active low) if spindle speed option not enabled
```
For this you can use fastio:

``` cpp
WRITE(SPINDLE_ON_OFF_PIN, SPINDLE_ON_OFF_INVERT)
```

Consider `SPINDLE_ENABLE_PIN` as an alternative name for this pin, and `SPINDLE_ENABLE_INVERT` for the option. Generally '`ENABLE`' is used for this kind of pin.

thinkyhead reviewed (changes requested):
Looks pretty good. Needs some updates. Suggested changes in comments.

I also have pending changes for `M3`, `M4`, `M5` laser support, but it will be a while before that's ready.

Bob-the-Kuhn commented:
1) fixed a bug when setting the maximum RPM

2) made it more generic - allowed the user to set the RPM to PWM
conversion constants in config_adv

3) lots of changes that Scott suggested

Changed “SPINDLE_ON_OFF” to “SPINDLE_ENABLE”

Changed “digitalWrite(” to “WRITE(“

Went to “#if ENABLED(SPINDLE_SPEED)” and ““#if ENABLED(SPINDLE_ENABLE)”

Removed the debug prints

Indented all the comments by 2 spaces

Got rid of the extra “pinMode” commands

Went to “gcode_M3_M4”

Went to “elseif” in sanitycheck.h

Bob-the-Kuhn commented:
I need your help on a couple of items.

I need to add a power up delay to allow the spindle speed to settle out.  As best I can tell the only way is to use the delay(xx) function and then to beat the dog using watchdog_reset().  I expected to find examples in the code but all I found were ones using PAUSE which allowed other commands to be processed.  Anything more elegant?

When I enter the commands manually through Repetier Host all is well.  When I have it execute a g-code file the spindle starts up but is almost immediately turned off when the G0 commands start executing.  Someone is setting my control pin to logic 1 (not tri-stated but driving 5V).  I've tried 4 pins with the same results.  I need suggestions on tracking down the issue.

Blue-Marlin commented:
If your program is very short, maybe a M400 right before shutting the spindel down again might help.
If the spindel needs time to accelerate, take a break and wait some time for it - G4 S123.

Bob-the-Kuhn commented:
That works.

  Thanks

PS - I swear I tried that earlier ...

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,58 @@
 #endif
+
+
+#if ENABLED( SPINDLE_ENABLE) && PIN_EXISTS(SPINDLE_ENABLE)
```
Better to have this stand alone as:

``` cpp
#if ENABLED(SPINDLE_ENABLE)
  #if !PIN_EXISTS(SPINDLE_ENABLE)
    #error "SPINDLE_ENABLE requires SPINDLE_ENABLE_PIN."
  #elif . . .
    . . .
  #endif
#endif // SPINDLE_ENABLE
```

thinkyhead reviewed (changes requested):
Almost there.... Still a little messy and has some bugs. Keep patching it up and it will get there.

Bob-the-Kuhn commented:
Made the change to sanitycheck.h.  That was a good idea.

Added user defined power up/down delays to allow the stepper speeds to stabilize.

Fixed my "mystery turn off" bug.  Just needed to add the synchronize() call.  I knew that some commands were done asynchronously to the movement commands, just hadn't realized that was the norm.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4279,6 +4282,144 @@ inline void gcode_G92() {
+
+#if ENABLED(SPINDLE_ENABLE)
+
+#define SPINDLE_POWER_UP    refresh_cmd_timeout();  \
```
Macros like this must be wrapped in `do{ ... }while(0)` and should not end with a semicolon.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4279,6 +4282,144 @@ inline void gcode_G92() {
+     if (spindle_speed == 0) { 
+         stepper.synchronize();   // wait until previous movement commands (G0, G1, G2 & G3) have completed before turning spindle off
+		 WRITE(SPINDLE_ENABLE_PIN, !SPINDLE_ENABLE_INVERT);  //turn spindle off (active low);
+         SPINDLE_POWER_DOWN
```
Macro invocations must end with a semicolon.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4279,6 +4282,144 @@ inline void gcode_G92() {
+     uint8_t spindle_speed = code_seen('S') ? code_value_byte() : 0; 
+     if (spindle_speed == 0) { 
+         stepper.synchronize();   // wait until previous movement commands (G0, G1, G2 & G3) have completed before turning spindle off
+		 WRITE(SPINDLE_ENABLE_PIN, !SPINDLE_ENABLE_INVERT);  //turn spindle off (active low);
```
2-space indentations only. No tabs!

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4279,6 +4282,144 @@ inline void gcode_G92() {
+         #if SPINDLE_SPEED_INVERT==true
+            spindle_speed = 255 - spindle_speed ;
+         #endif         
+         analogWrite(SPINDLE_SPEED_PIN,spindle_speed);
```
Space after comma.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -7413,6 +7554,20 @@ void process_next_command() {
       #endif // ULTIPANEL
+     
+      #if ENABLED(SPINDLE_ENABLE)
+        case 3:
```
Note that other GCode commands include a brief synopsis on the `case N:` line. Feel free to add yours here.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -9421,6 +9576,18 @@ void setup() {
+     OUT_WRITE(SPINDLE_ENABLE_PIN, !SPINDLE_ENABLE_INVERT);  //init spindle to off;
+     #if ENABLED(SPINDLE_SPEED)
+       pinMode(SPINDLE_SPEED_PIN, OUTPUT);
+       #if SPINDLE_SPEED_INVERT == false        // init PWM to lowset speed
```
Since you've chosen to make `SPINDLE_SPEED_INVERT` a `true` or `false` value you should use it with that in mind. So instead of conditional block here, use it like this instead:

``` cpp
analogWrite(SPINDLE_SPEED_PIN, SPINDLE_SPEED_INVERT ? 255 : 0);
```

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,59 @@
 #endif
+
+
+#if ENABLED( SPINDLE_ENABLE)
```
Spacing in `SanityCheck.h` still needs a conscientious cleanup.
It looks like it was typed out by a dolphin wearing mittens.

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -983,3 +983,59 @@
+  #elif PIN_EXISTS(CONTROLLERFAN) && SPINDLE_SPEED_PIN == CONTROLLERFAN_PIN
+     #error "SPINDLE_SPEED PWM pin is used by CONTROLLERFAN_PIN "
+  #elif defined(HAS_MOTOR_CURRENT_PWM) && \
+        ((PIN_EXISTS(MOTOR_CURRENT_PWM_XY) &&  SPINDLE_SPEED_PIN == MOTOR_CURRENT_PWM_ ) \
```
You probably meant `MOTOR_CURRENT_PWM_XY` here and not `MOTOR_CURRENT_PWM_`.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4279,6 +4282,144 @@ inline void gcode_G92() {
+                            if (SPINDLE_POWER_UP_DELAY > 0)  while (PENDING(millis(), SPINDLE_POWER_UP_DELAY * 1000 + previous_cmd_ms)) idle(); \ 
+                            else while (PENDING(millis(), 1000 + previous_cmd_ms)) idle();                                                       // allow at least 1 second for power up
+
+#define SPINDLE_POWER_DOWN  refresh_cmd_timeout();  \
```
And actually, there's no special reason why these need to be macros. Why not just make these into sub-functions instead?

thinkyhead commented on Marlin/Configuration_adv.h:
```diff
@@ -804,4 +804,114 @@
+
+
+
+
```
As special as this feature is, there's no need for quite so many blank lines here. Remember, whatever is added here needs to be added to all other `Configuration_adv.h` files.

thinkyhead commented on Marlin/Configuration_adv.h:
```diff
@@ -804,4 +804,114 @@
+//#define SPINDLE_SPEED
+ 
+ /*
+ *
```
Please pay attention to comment formatting. There's a space before the lines starting with asterisk, so the asterisks align, and the first line contains a "Doxygen-style" extra asterisk (for future use with Doxygen).

``` cpp
/**
 *
 *
 *
 */
```

thinkyhead commented on Marlin/Configuration_adv.h:
```diff
@@ -804,4 +804,114 @@
+ #define RPM_slope 302.00    // SuperPID & AzteegX3pro controller pin 5
+ #define RPM_intercept 0     // SuperPID & AzteegX3pro controller pin 5
+ #define RPM_min  5000       // SuperPID
+ #define RPM_max 30000       // SuperPID
```
Configuration values created using `#define` should be all-caps.

``` cpp

#define RPM_SLOPE 302.00    // SuperPID & AzteegX3pro controller pin 5
#define RPM_INTERCEPT 0     // SuperPID & AzteegX3pro controller pin 5
#define RPM_MIN  5000       // SuperPID
#define RPM_MAX 30000       // SuperPID
```

Bob-the-Kuhn replied:
These are all implemented in the latest commit

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4279,6 +4282,144 @@ inline void gcode_G92() {
+
+
+ 
+ 
```
Fewer blank lines, please.

thinkyhead reviewed (changes requested):
Please make the requested changes. And please use the following procedure to rebase to the latest code and to squash all excess commits:

```
git remote add upstream https://github.com/MarlinFirmware/Marlin.git
git fetch upstream
git rebase upstream/RCBugFix
git rebase -i upstream/RCBugFix
git push -f
```

Bob-the-Kuhn commented on Marlin/Configuration_adv.h:
```diff
@@ -804,4 +804,114 @@
+ #define RPM_slope 302.00    // SuperPID & AzteegX3pro controller pin 5
+ #define RPM_intercept 0     // SuperPID & AzteegX3pro controller pin 5
+ #define RPM_min  5000       // SuperPID
+ #define RPM_max 30000       // SuperPID
```
These are all implemented in the latest commit

Bob-the-Kuhn committed:
fix conflict

compared against the latest config_adv in RCBugfix and added the newer
lines (M43 stuff)

Bob-the-Kuhn commented:
Today's commit should take care of the "Conflicting files" issue.

Bob-the-Kuhn commented:
My repository is gone so I can't resolve the conflicts and push them to this PR.

The best I can do is add the updated files to this comment.

[Configuration_adv.zip](https://github.com/MarlinFirmware/Marlin/files/559626/Configuration_adv.zip)
[Marlin_main.zip](https://github.com/MarlinFirmware/Marlin/files/559625/Marlin_main.zip)

thinkyhead commented:
> My repository is gone

How'd that happen? Can you not restore it and make a new PR?

Bob-the-Kuhn commented:
I deleted the repository (so I could work on the M43 end stop monitor) because I didn't understand that I could work multiple local branches independently of each other and GitHub would magically keep them separate even though there is only one directory on my PC.

I'll create a new branch, copy over the modified files and do a new PR.

Bob-the-Kuhn commented:
This PR has been replaced by PR #5113.

PR #5113 is a clone of this PR with the conflicts resolved.

There were no changes to the spindle specific sections.

thinkyhead commented:
@Bob-the-Kuhn This is awesome, and taught me a lot! http://learngitbranching.js.org/

Bob-the-Kuhn commented:
Thanks for the link.

I finally figured out how to properly work on two issues at the same time.
